crates/solidity/outputs/cargo/crate/src/extensions/utils/version/mod.rs

```diff
@@ -0,0 +1,458 @@
+        ComparatorSet {
+            comparators: vec![comp],
+        }
+    }
+
+    fn bounds(lower: Comparator, upper: Comparator) -> ComparatorSet {
+        ComparatorSet {
+            comparators: vec![lower, upper],
+        }
+    }
+
+    fn wild() -> ComparatorSet {
+        ComparatorSet::single(Comparator::wild())
+    }
+
+    fn none() -> ComparatorSet {
+        ComparatorSet::single(Comparator::none())
+    }
+
+    /// Create a `ComparatorSet` which represents the range of versions described by `^[partial]`.
+    fn caret(partial: &PartialVersion) -> ComparatorSet {
+        match partial.caret_upper_bound() {
+            Some(upper_version) => {
+                let upper = Comparator {
+                    version: upper_version,
+                    op: Operator::Lt,
+                };
+
+                let lower = Comparator {
+                    version: partial.lower_bound(),
+                    op: Operator::GtEq,
+                };
+
+                ComparatorSet::bounds(lower, upper)
+            }
+            None => ComparatorSet::wild(),
+        }
+    }
+
+    /// Create a `ComparatorSet` which represents the range of versions described by `~[partial]`.
+    fn tilde(partial: &PartialVersion) -> ComparatorSet {
+        match partial.tilde_upper_bound() {
+            Some(upper_version) => {
+                let upper = Comparator {
+                    version: upper_version,
+                    op: Operator::Lt,
+                };
+
+                let lower = Comparator {
+                    version: partial.lower_bound(),
+                    op: Operator::GtEq,
+                };
+
+                ComparatorSet::bounds(lower, upper)
+            }
+            None => ComparatorSet::wild(),
+        }
+    }
+
+    /// Create a `ComparatorSet` which represents the range of versions described by `[lower_version] - [upper_version]`.
+    fn hyphen_range(
+        lower_version: &PartialVersion,
+        upper_version: &PartialVersion,
+    ) -> ComparatorSet {
+        let lower = Comparator {
+            version: lower_version.lower_bound(),
+            op: Operator::GtEq,
+        };
+
+        let upper = match upper_version.major {
+            VersionPart::Specified(major) => match upper_version.minor {
+                VersionPart::Specified(minor) => match upper_version.patch {
+                    VersionPart::Specified(patch) => Comparator {
+                        version: Version::new(major, minor, patch),
+                        op: Operator::LtEq,
+                    },
+                    _ => Comparator {
+                        version: Version::new(major, minor + 1, 0),
+                        op: Operator::Lt,
+                    },
+                },
+                _ => Comparator {
+                    version: Version::new(major + 1, 0, 0),
+                    op: Operator::Lt,
+                },
+            },
+            _ => Comparator::wild(),
+        };
```

Comment:
In terms of the upper bound functions in `PartialVersion` I think this can be written as:
```
let upper = if upper_version.is_complete() {
  Comparator {
    version: upper_version.lower_bound(),
    op: Operator::LtEq,
  }
} else {
  Comparator {
    version: upper_version.tilde_upper_bound(),
    op: Operator::Lt,
  }
};
```